crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,4433 @@
+    TokenKind::Import,
+    TokenKind::From,
+    TokenKind::Type,
+]);
+/// Tokens that represent simple statements, including expressions.
+const SIMPLE_STMT_SET2: TokenSet = SIMPLE_STMT_SET.union(EXPR_SET);
+
+impl<'src, I> Parser<'src, I>
+where
+    I: Iterator<Item = LexResult>,
+{
+    pub(crate) fn new(
+        source: &'src str,
+        mode: Mode,
+        lexer: I,
+    ) -> Parser<'src, impl Iterator<Item = LexResult>> {
+        Parser {
+            mode,
+            source,
+            errors: Vec::new(),
+            ctx_stack: Vec::new(),
+            ctx: ParserCtxFlags::empty(),
+            last_ctx: ParserCtxFlags::empty(),
+            lexer: itertools::peek_nth(lexer),
+        }
+    }
+
+    pub(crate) fn parse(mut self) -> ParsedFile {
+        let mut body = vec![];
+
+        let ast = if self.mode == Mode::Expression {
+            let (expr, range) = self.parse_exprs();
+            loop {
+                if !self.eat(TokenKind::Newline) {
+                    break;
+                }
+            }
+            self.expect(TokenKind::EndOfFile);
+
+            ast::Mod::Expression(ast::ModExpression {
+                body: Box::new(expr),
+                range,
+            })
+        } else {
+            let is_src_empty = self.at(TokenKind::EndOfFile);
+            while !self.at(TokenKind::EndOfFile) {
+                if self.at(TokenKind::Indent) {
+                    self.handle_unexpected_indentation(&mut body, "unexpected indentation");
+                    continue;
+                }
+                let (stmt, _) = self.parse_statement();
+                body.push(stmt);
+            }
+            ast::Mod::Module(ast::ModModule {
+                body,
+                // If the `source` only contains comments or empty spaces, return
+                // an empty range.
+                range: if is_src_empty {
+                    TextRange::default()
+                } else {
+                    TextRange::new(
+                        0.into(),
+                        self.source
+                            .len()
+                            .try_into()
+                            .expect("source length is  bigger than u32 max"),
+                    )
+                },
+            })
+        };
+
+        // After parsing, the `ctx` and `ctx_stack` should be empty.
+        // If it's not, you probably forgot to call `clear_ctx` somewhere.
+        assert!(self.ctx.is_empty() && self.ctx_stack.is_empty());
+
+        ParsedFile {
+            ast,
+            parse_errors: self.errors,
+        }
+    }
+
+    #[inline]
+    fn set_ctx(&mut self, ctx: ParserCtxFlags) {
+        self.ctx_stack.push(self.ctx);
+        self.ctx = ctx;
+    }
+
+    #[inline]
+    fn clear_ctx(&mut self, ctx: ParserCtxFlags) {
+        assert_eq!(self.ctx, ctx);
+        self.last_ctx = ctx;
+        if let Some(top) = self.ctx_stack.pop() {
+            self.ctx = top;
+        }
```

Comment:
`self.ctx_stack` will only be empty at the end of the parsing, if it's empty during the parsing, `self.set_ctx` or `self.clear_ctx` got misused somehow. To avoid this, I think we need to add an `assert!(!self.ctx_stack.is_empty())` in `self.clear_ctx`.